File: rust/src/code/spiral_matrix.rs
pub struct Solution {}

impl Solution {
    pub fn spiral_order(matrix: Vec<Vec<i32>>) -> Vec<i32> {
        let mut result = Vec::<i32>::new();
        let width = matrix[0].len();
        let height = matrix.len();
        let mut x_s = 0;
        let mut x_e = width - 1;
        let mut y_s = 0;
        let mut y_e = height - 1;

        while x_s <= x_e && y_s <= y_e {
            // left -> right
            for i in x_s..=x_e {
                result.push(matrix[y_s][i])
            }
            y_s += 1;

            // up -> down
            for i in y_s..=y_e {
                result.push(matrix[i][x_e])
            }
            if let None = x_e.checked_sub(1) {
                break;
            }
            x_e -= 1;

            // right -> left
            if y_s <= y_e {
                for i in (x_s..=x_e).rev() {
                    result.push(matrix[y_e][i])
                }
            }
            if let None = y_e.checked_sub(1) {
                break;
            }
            y_e -= 1;

            // down -> up
            if x_s <= x_e {
                for i in (y_s..=y_e).rev() {
                    result.push(matrix[i][x_s])
                }
            }
            x_s += 1;
        }

        result
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn case_1() {
        // arrange
        let matrix = vec![vec![1, 2, 3], vec![4, 5, 6], vec![7, 8, 9]];
        let expected = vec![1, 2, 3, 6, 9, 8, 7, 4, 5];

        // act
        let actual = Solution::spiral_order(matrix);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn case_2() {
        // arrange
        let matrix = vec![vec![1, 2, 3, 4], vec![5, 6, 7, 8], vec![9, 10, 11, 12]];
        let expected = vec![1, 2, 3, 4, 8, 12, 11, 10, 9, 5, 6, 7];

        // act
        let actual = Solution::spiral_order(matrix);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn case_3() {
        // arrange
        let matrix = vec![vec![1]];
        let expected = vec![1];

        // act
        let actual = Solution::spiral_order(matrix);

        // assert
        assert_eq!(expected, actual);
    }
}

File: python/spiral_matrix.py
class Solution:
    def spiralOrder(self, matrix: list[list[int]]) -> list[int]:
        """time complexity: O(n * m)"""
        result: list[int] = []
        s_r: int = 0
        e_r: int = len(matrix) - 1
        s_c: int = 0
        e_c: int = len(matrix[0]) - 1
        while s_r <= e_r and s_c <= e_c:
            # left -> right
            for i in range(s_c, e_c + 1):
                result.append(matrix[s_r][i])
            s_r += 1
            # up -> down
            for i in range(s_r, e_r + 1):
                result.append(matrix[i][e_c])
            e_c -= 1
            if s_r <= e_r:
                # right -> left
                for i in range(e_c, s_c - 1, -1):
                    result.append(matrix[e_r][i])
                e_r -= 1
            if s_c <= e_c:
                # down -> up
                for i in range(e_r, s_r - 1, -1):
                    result.append(matrix[i][s_c])
                s_c += 1
        return result


def test_spiralOrder_case_1():
    # arragne
    matrix: list[list[int]] = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
    expected: list[int] = [1, 2, 3, 6, 9, 8, 7, 4, 5]

    # act
    actual: list[int] = Solution().spiralOrder(matrix)

    # assert
    assert expected == actual


def test_spiralOrder_case_2():
    # arragne
    matrix: list[list[int]] = [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]
    expected: list[int] = [1, 2, 3, 4, 8, 12, 11, 10, 9, 5, 6, 7]

    # act
    actual: list[int] = Solution().spiralOrder(matrix)

    # assert
    assert expected == actual


def test_spiralOrder_case_3():
    # arragne
    matrix: list[list[int]] = [[2, 5, 8], [4, 0, -1]]
    expected: list[int] = [2, 5, 8, -1, 0, 4]

    # act
    actual: list[int] = Solution().spiralOrder(matrix)

    # assert
    assert expected == actual
